Arbitrary binning axes

steinnymir opened the pull request:
Implementing binning with arbitrary axes, allowing for non-linear binnings.

It works for bin_partition, but still not for bin_dataframe.
The reason is I am trying to use nBins as either the number of bins in the old method, but also to pass it arrays directly.
This then drops the requirement of binRanges. I will try fix this....

However the numba function works as of now, even if I had to use a workaround for the creation of the target nd-array.
I always hit this problem where you cannot create tuples in Numba.

@daviddoji if you can have a look at this would be great! the tutorial notebook has tests which use this new function.

daviddoji commented on sed/binning.py:
```diff
@@ -199,8 +199,8 @@ def bin_dataframe(
+                        binRanges,
                         hist_mode,
                     ),
                 )
```
Is there any good reason to keep `legacy` and `lean` modes, apart from compatibility?

steinnymir replied:
They were there mostly to test which one performs best. For the time being I would leave them

daviddoji commented on sed/binning.py:
```diff
@@ -355,7 +355,85 @@ def _hist_from_bin_ranges(
             j = (sample[t, i] - ranges[i, 0]) * delta[i]
             is_inside = is_inside and (0 <= j < bins[i])
             flatidx += int(j) * strides[i]
+            if not is_inside:
```
I'd put comments before the statements. In this case, before the `if not is_inside`

daviddoji commented on sed/binning.py:
```diff
@@ -355,7 +355,85 @@ def _hist_from_bin_ranges(
+
+
+@numba.jit(nogil=True, parallel=False, nopython=True)
+def binsearch(bins: np.ndarray, val: float) -> int:
```
is this similar to the already implemented `bisect_left` from the standard library (https://docs.python.org/3/library/bisect.html)?

steinnymir replied:
yep, looks exactly the same! however:

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'bisect_left': Cannot determine Numba type of <class 'builtin_function_or_method'>

File "..\sed\binning.py", line 495:
def _hist_from_bins(
    <source elided>
            # j = binsearch(bins[i], sample[t, i])
            j = bisect_left(bins[i], sample[t, i])
            ^

Numba doesn't know bisect_left, so I had to do it manually.

daviddoji replied:
yeah, numba is a bit picky with that. You can try this one from numpy:
https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html

According to numba's docs, it's supported:
https://numba.readthedocs.io/en/stable/reference/numpysupported.html?highlight=numpy%20

Sorry for being harsh on this, but my point on this is to reuse well-tested functions instead of creating them by ourselves :)

daviddoji commented on sed/binning.py:
```diff
@@ -355,7 +355,85 @@ def _hist_from_bin_ranges(
+    Hflat = H.ravel()
+
+    strides = np.zeros(ndims, np.int64)
+    # print(H.shape,strides.shape,H.strides.shape,H.itemsize)
```
Do not forget to remove commented lines used for debugging :)

daviddoji commented on sed/binning.py:
```diff
@@ -364,9 +442,9 @@ def _hist_from_bin_ranges(
+    bins: Union[Sequence[Union[int, np.ndarray]], np.ndarray],
+    ranges: Sequence = None,
+) -> Tuple[np.ndarray, np.ndarray]:
     """Wrapper for the Number pre-compiled binning functions.
```
I would skip the reason why the function was created. The user do not care :)
Regarding the overflow, I would use an assert for that

steinnymir replied:
good point! I will look into the function descriptions again

daviddoji commented on sed/binning.py:
```diff
@@ -405,37 +483,56 @@ def numba_histogramdd(
+            method = "int"
+        elif isinstance(bins[0], np.ndarray):
+            method = "array"
+
```
remove empty line

daviddoji commented on sed/binning.py:
```diff
@@ -405,37 +483,56 @@ def numba_histogramdd(
-    except TypeError:
-        # bins is an integer
+        Db = len(bins)
+        if isinstance(bins[0], int):
```
would be simpler if `bins` are always either `ints` or `arrays`??

steinnymir replied:
Not sure what you mean...

daviddoji replied:
why is it necessary to support either bins or arrays?
Would not be easier to support only arrays?

steinnymir replied:
we have two binning methods now, implemented with numba. One based on defining the number of bins (therefore int) and is much faster. The other takes a pre-made array and is slower, but allows non-linear binning.

Therefore the two different input methods. Also, this mimics what numpy.histogramdd does, where you can pass an int or an array.

daviddoji reviewed (commented):
Apart from the comments, I'm also missing tests for the functions

steinnymir commented on sed/binning.py:
```diff
@@ -28,8 +28,8 @@ def bin_partition(
     part: Union[dask.dataframe.core.DataFrame, pd.DataFrame],
     binDict: dict = None,
```
removed!

daviddoji replied:
My apologies @steinnymir. I meant to use snake case, as it is the standard in the Python community

steinnymir commented:
I changed the function argument names to better follow the numpy naming in similar functions.
Also added multiple ways to define the bins:
- an integer as before
- a tuple
- a pre-defined numpy array
all the above can be passed directly or as a list of them.

Finally, one can also pass a dictionary with keys the names of the columns and values being any of the above.
For now, the same type needs  to be used for all the dimensions on which we are binning, i.e. no mixing between the three types above.

I started making a test function, but have not "tested" it yet. I will complete the pull request once all functionalities have a test function. This is the right approach right?

steinnymir commented:
The test functions I made don't work. I guess I need to learn more about testing.... @daviddoji could you have a look and tell me what I am doing wrong there? 😓 

I tested it in the tutorial, which changed accordingly.

Meanwhile I added also the jitter functionality, so now the binning features should be complete!
This would close #26

rettigl commented:
I played around a bit with this. The version that uses the bin edges array as input does not produce the same output as the version with start and enpoints and number of bins. Specifically, there seems to be a problem with the last bin, which contained 0 in my test in the version using the bin edges.
On the plus side, I did not observe any performance differences between the two approaches (however, still with equidistant bins).
Also, just passing a list of the axes on which to apply jittering does not do anything at the moment, seems this is not implemented. I will have a look at the version with the dictionary.

daviddoji commented:
is this PR ready for merging? @steinnymir

steinnymir commented:
> The version that uses the bin edges array as input does not produce the same output as the version with start and enpoints and number of bins

@rettigl I also observed this once, but I cannot reproduce it any more. Looking at real data, binned with both methods and with the hextof library, all give the same exact result.

Therefore for now I merge this pull request, and if this problem reappears, we can add an issue for it.
